src/gpu/amd/miopen_softmax_impl.hpp

```diff
@@ -0,0 +1,269 @@
+/*******************************************************************************
+* Copyright 2020-2022 Intel Corporation
+* Copyright 2020 Codeplay Software Limited 
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_AMD_MIOPEN_SOFTMAX_IMPL_HPP
+#define GPU_AMD_MIOPEN_SOFTMAX_IMPL_HPP
+
+#include "gpu/amd/sycl_hip_utils.hpp"
+#include "miopen/miopen.h"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace amd {
+
+struct miopen_softmax_impl_base_t {
+    enum io { src = 0, dst, d_src, d_dst, NUM_IO };
+    int strides[NUM_IO][DNNL_MAX_NDIMS];
+    miopenDataType_t data_type;
+    int ndims;
+    miopenSoftmaxAlgorithm_t alg_kind;
+
+    miopenSoftmaxMode_t mode = miopenSoftmaxMode_t::MIOPEN_SOFTMAX_MODE_CHANNEL;
+
+    // oneDNN softmax primitive doesn't support any post-ops or attributes,
+    // hence we can set alpha = 1 and beta = 0 for all cases
+    float alpha = 1.0f;
+    float beta = 0.0f;
+
+    virtual ~miopen_softmax_impl_base_t() {}
+
+    virtual status_t init(const softmax_pd_t *pd) = 0;
+
+    virtual void execute(miopenHandle_t handle, void **x, int size) const = 0;
+
+    // Mapping between dnnl algorithm and cuDNN softmax algorithm
+    status_t convert_alg_kind(bool is_log_softmax,
+            miopenSoftmaxAlgorithm_t *miopen_alg_kind) const {
+        if (is_log_softmax) {
+            *miopen_alg_kind = miopenSoftmaxAlgorithm_t::MIOPEN_SOFTMAX_LOG;
+        } else {
+            *miopen_alg_kind
+                    = miopenSoftmaxAlgorithm_t::MIOPEN_SOFTMAX_ACCURATE;
+        }
+
+        return status::success;
+    }
+
+    status_t convert_dims_softmax(const dims_t &orig_dims, int *modified_dims,
+            int axis, int ndims, format_tag_t tag,
+            miopenTensorFormat_t &format) const {
+
+        // Initialise all dims to 1
+        for (int i = 0; i < 4; i++) {
+            modified_dims[i] = 1;
+        }
+        if (axis == 1) {
```

Comment:
If this primitive only supports `axis = 1` then why do we need the code below for the case when `axis != 1`?

Reply:
Modified the logic to handle only nchw format.